generators/ruby-v2/ast/src/ast/Type.ts

```diff
@@ -1,46 +1,341 @@
 import { assertNever } from "@fern-api/core-utils";
 
+import { TypeParameter } from "./TypeParameter";
 import { AstNode } from "./core/AstNode";
 import { Writer } from "./core/Writer";
 
-type InternalType = Boolean_ | String_;
+interface Self {
+    type: "self";
+}
+
+interface Class_ {
+    type: "class";
+}
+
+interface Instance {
+    type: "instance";
+}
 
 interface Boolean_ {
     type: "boolean";
 }
 
+interface Nil {
+    type: "nil";
+}
+
+interface Top {
+    type: "top";
+}
+
+interface Bot {
+    type: "bot";
+}
+
+interface Void {
+    type: "void";
+}
+
+interface Boolish {
+    type: "boolish";
+}
+
 interface String_ {
     type: "string";
 }
 
+interface Integer {
+    type: "integer";
+}
+
+interface Union {
+    type: "union";
+    elems: Type[];
+}
+
+interface Intersection {
+    type: "intersection";
+    elems: Type[];
+}
+
+interface Array_ {
+    type: "array";
+    elem: Type;
+}
+
+interface Hash {
+    type: "hash";
+    keyType: Type;
+    valueType: Type;
+}
+
+interface Object_ {
+    type: "object";
+    klass: string;
+}
+
+interface Singleton {
+    type: "singleton";
+    klass: string;
+}
+
+interface Tuple {
+    type: "tuple";
+    elems: Type[];
+}
+
+interface Generic {
+    type: "generic";
+    baseKlass: string;
+    parameters: (Type | TypeParameter)[];
+}
+
+export type BaseType = Self | Class_ | Instance | Boolean_ | Nil | Top | Bot | Void;
+export type SingleType = Boolish | String_ | Integer | Union | Intersection | Singleton | Object_ | Generic;
+// TODO: These conceivably could just be Generics, at the cost of some specificity/usability
```

Comment:
i like them being specific as you have them since they are likely going to be used more often